File: web/api/summarize.go
package api

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"gonotes/summarize"

	"github.com/rohanthewiz/logger"
	"github.com/rohanthewiz/rweb"
)

// ============================================================================
// Summarize API
//
// One endpoint, two callers: the web UI's Summarize buttons and the TUI's HTTP
// store (a TUI in local mode calls the summarize package directly and never
// comes through here).
//
// WHY THE SERVER DOES THIS AND NOT THE BROWSER: the summarizer is the local
// `claude` CLI (see package summarize), which exists on the machine running
// GoNotes, not in the tab. That also means the answer to "can I summarize?"
// is a property of the SERVER — so GET reports it, and the web UI hides its
// buttons rather than offering a door that always fails.
//
// Authentication is required even though nothing is read from or written to
// the database: this endpoint spends money and CPU on the host's account, so
// it must not be reachable by anyone who can reach the port.
// ============================================================================

// summarizeTimeout bounds the model call. Generous, because the input may be a
// long paste on a slow link, but finite: without it a wedged CLI would hold the
// handler — and the browser's spinner — open forever.
const summarizeTimeout = 3 * time.Minute

// SummarizeStatus handles GET /api/v1/summarize
// Reports whether this installation can summarize at all, so a UI can decide
// whether to show the feature. Never an error: "no" is a legitimate answer.
func SummarizeStatus(ctx rweb.Context) error {
	if GetCurrentUserGUID(ctx) == "" {
		return writeError(ctx, http.StatusUnauthorized, "authentication required")
	}
	return writeSuccess(ctx, http.StatusOK, map[string]any{
		"available":     summarize.Available(),
		"default_model": summarize.DefaultModel,
	})
}

// Summarize handles POST /api/v1/summarize
// Request body: {"text": "...", "model": "haiku"}  (model optional)
// Response data: {"title": "...", "description": "...", "body": "..."}
//
// Nothing is stored. The caller gets the three fields and decides what to do
// with them — which is what lets the same endpoint serve "summarize the
// clipboard into a new note" and "summarize what is in the editor right now".
func Summarize(ctx rweb.Context) error {
	userGUID := GetCurrentUserGUID(ctx)
	if userGUID == "" {
		return writeError(ctx, http.StatusUnauthorized, "authentication required")
	}

	var req struct {
		Text  string `json:"text"`
		Model string `json:"model"`
	}
	if err := json.Unmarshal(ctx.Request().Body(), &req); err != nil {
		return writeError(ctx, http.StatusBadRequest, "invalid request body")
	}
	// Trimmed, not just empty: a body of spaces and newlines is a client mistake
	// (an empty editor, a clipboard holding a blank line), and it deserves a 400
	// saying so rather than the 502 it would earn by reaching the summarizer and
	// being refused there.
	if strings.TrimSpace(req.Text) == "" {
		return writeError(ctx, http.StatusBadRequest, "nothing to summarize")
	}

	// 503 rather than 500 for a missing CLI: the request was fine, the capability
	// is absent, and that distinction is what tells the UI to explain rather than
	// to apologize.
	if !summarize.Available() {
		return writeError(ctx, http.StatusServiceUnavailable,
			"summarizing needs the 'claude' CLI on this machine's PATH")
	}

	cCtx, cancel := context.WithTimeout(context.Background(), summarizeTimeout)
	defer cancel()

	res, err := summarize.Summarize(cCtx, req.Text, req.Model)
	if err != nil {
		// Logged here and reported to the caller as its own message: a failed
		// summary is usually explained by what the model said back (see
		// summarize.parseResult), and hiding that behind a generic string would
		// leave the user with nothing to act on.
		//
		// The note text itself is NEVER logged — only the failure. Bodies do not
		// go to disk outside the database.
		logger.LogErr(err, "summarize failed", "user", userGUID)
		return writeError(ctx, http.StatusBadGateway, err.Error())
	}

	return writeSuccess(ctx, http.StatusOK, res)
}

File: web/static/js/summarize.js
// summarize.js — the web UI's two doors into the summarizer.
//
//   toolbar     Summarize clipboard → a NEW note, prefilled and unsaved
//   edit footer Summarize           → replaces the body being edited, in place
//
// Both post to /api/v1/summarize, which runs the `claude` CLI on the machine
// the SERVER runs on (see web/api/summarize.go and package summarize). Two
// consequences shape this file:
//
//   - WHETHER THE FEATURE EXISTS IS A SERVER PROPERTY. A host without the CLI
//     cannot summarize no matter what the browser can do, so both buttons ship
//     hidden in the HTML and are revealed only after the GET says "available".
//     Revealing rather than hiding avoids a button that flashes on every load.
//   - THE CLIPBOARD IS A BROWSER PROPERTY. navigator.clipboard.readText needs a
//     secure context (localhost counts; a plain-http LAN address does not) and
//     a permission the user may refuse. That failure is not the summarizer's,
//     and is reported as its own thing with a usable fallback.
//
// Nothing here saves. A summary lands in the form and waits for Save — the same
// rule the TUI keeps (tui/summarize.go), and for the same reason: what a model
// wrote about a text is a proposal, and the person who pasted the text is the
// one who decides whether it is worth keeping.
(function() {
  'use strict';

  window.app = window.app || {};

  // Both buttons, resolved lazily: this file loads with the page, but the probe
  // that reveals them runs later.
  function clipBtn() { return document.getElementById('btn-summarize-clip'); }
  function bodyBtn() { return document.getElementById('btn-summarize-body'); }

  function internals() { return (window.app && window.app._internal) || null; }

  function toast(message, kind) {
    const api = internals();
    if (api && api.showToast) api.showToast(message, kind || 'info');
    else console.log('[summarize]', message);
  }

  // ============================================
  // Availability probe
  // ============================================

  // Asked once per page load. A failure (offline, 401 on the way to the login
  // redirect) leaves the buttons hidden, which is the safe direction: a hidden
  // button costs a user who has the CLI one reload, whereas a visible one on a
  // server without it costs every click.
  async function probeAvailability() {
    const api = internals();
    if (!api) return;
    try {
      const res = await api.apiRequest('/summarize');
      if (res && res.data && res.data.available) {
        if (clipBtn()) clipBtn().style.display = '';
        if (bodyBtn()) bodyBtn().style.display = '';
      }
    } catch (err) {
      console.warn('summarize: availability probe failed', err);
    }
  }

  // ============================================
  // The call
  // ============================================

  // requestSummary posts text and returns {title, description, body}, or null
  // after reporting why not. The server's own message is what gets shown: it
  // names what the model said back, or that the CLI is missing, and a generic
  // "summarize failed" would leave the user with nothing to act on.
  async function requestSummary(text) {
    const api = internals();
    if (!api) return null;
    try {
      const res = await api.apiRequest('/summarize', {
        method: 'POST',
        body: JSON.stringify({ text: text })
      });
      if (!res || !res.data || !res.data.body) {
        toast('The summarizer returned nothing usable', 'error');
        return null;
      }
      return res.data;
    } catch (err) {
      // No toast here: apiRequest already showed the server's message on its
      // way out, and a second one saying the same thing reads as two failures.
      console.warn('summarize: request failed', err);
      return null;
    }
  }

  // busy disables a button and says what it is doing, so a second click cannot
  // start a second model call over the same text. The label is restored from
  // what was there, not from a literal, so this cannot rename a button.
  function busy(btn, on, label) {
    if (!btn) return;
    if (on) {
      btn.dataset.prevLabel = btn.dataset.prevLabel || btn.textContent;
      btn.disabled = true;
      if (label && btn.textContent.trim()) btn.textContent = label;
    } else {
      btn.disabled = false;
      if (btn.dataset.prevLabel) {
        btn.textContent = btn.dataset.prevLabel;
        delete btn.dataset.prevLabel;
      }
    }
  }

  // ============================================
  // Body helpers
  // ============================================
  //
  // The body lives in the textarea; Monaco, when active, is a second surface
  // over the same value. Reading goes through _syncMonacoToBody and writing
  // through _syncBodyToMonaco, because assigning textarea.value fires no event
  // the editor could notice.

  function readBody() {
    if (window.app._syncMonacoToBody) window.app._syncMonacoToBody();
    const ta = document.getElementById('edit-body');
    return ta ? ta.value : '';
  }

  function writeBody(text) {
    const ta = document.getElementById('edit-body');
    if (ta) ta.value = text;
    if (window.app._syncBodyToMonaco) window.app._syncBodyToMonaco();
  }

  // ============================================
  // Door 1 — the clipboard
  // ============================================

  window.app.summarizeClipboard = async function() {
    const btn = clipBtn();

    let text = '';
    try {
      text = await navigator.clipboard.readText();
    } catch (err) {
      // A refused or unavailable clipboard is not a failure of this feature,
      // and the useful answer is the way around it: open an empty note, let the
      // user paste with ⌘V, and summarize that with the other button.
      toast('The browser would not give up the clipboard — paste into a new note and use Summarize', 'error');
      if (window.app.newNote) window.app.newNote();
      return;
    }

    if (!text || !text.trim()) {
      toast('Nothing to summarize — the clipboard is empty', 'info');
      return;
    }

    // The toolbar button is icon-only, so busy() has nothing to relabel — all
    // it can show is the disabled state (.btn:disabled in app.css). The toast
    // carries the rest, and says the same thing the TUI's status line says on
    // ctrl+r. It auto-dismisses after 3s and a model call can outlast that;
    // past then the dimmed button is what tells the user the wait is real.
    toast('Summarizing the clipboard…', 'info');
    busy(btn, true);
    const res = await requestSummary(text);
    busy(btn, false);
    if (!res) return;

    // A NEW note, deliberately: the clipboard door never touches an existing
    // one. newNote() clears the form and mints the guid, so everything below is
    // filling in blanks.
    window.app.newNote();
    const title = document.getElementById('edit-title');
    const desc = document.getElementById('edit-description');
    if (title) title.value = res.title || '';
    if (desc) desc.value = res.description || '';
    writeBody(res.body);
    if (title) title.focus();

    toast('Summarized from the clipboard — Save to keep it', 'success');
  };

  // ============================================
  // Door 2 — the body on screen
  // ============================================

  window.app.summarizeBody = async function() {
    const btn = bodyBtn();
    const text = readBody();
    if (!text || !text.trim()) {
      toast('Nothing to summarize — the body is empty', 'info');
      return;
    }

    // Announced as well as relabelled: the footer button can be scrolled out of
    // view or simply not where the eye is after a click, and this door is the
    // one that REPLACES text the user wrote. Being told it started is what
    // makes the replacement, when it lands, an expected event.
    toast('Summarizing the note body…', 'info');
    busy(btn, true, 'Summarizing…');
    const res = await requestSummary(text);
    busy(btn, false);
    if (!res) return;

    // The body is REPLACED — that is the request — but the note on disk is
    // untouched until Save, and Cancel still walks away from all of it. The
    // toast says so, because a replaced body is the one change here the user
    // did not watch happen.
    writeBody(res.body);

    // Title and description are only FILLED IN, never overwritten: a title the
    // user typed is their own words, and this request was only about condensing
    // the text. Same rule as gn-clip.sh's -t and the TUI's applySummary.
    const filled = [];
    const title = document.getElementById('edit-title');
    const desc = document.getElementById('edit-description');
    if (title && !title.value.trim() && res.title) { title.value = res.title; filled.push('title'); }
    if (desc && !desc.value.trim() && res.description) { desc.value = res.description; filled.push('description'); }

    const suffix = filled.length ? ' (' + filled.join(' and ') + ' filled in)' : '';
    toast('Body replaced with the summary' + suffix + ' — Save to keep it', 'success');
  };

  // ============================================
  // Wiring
  // ============================================

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', probeAvailability);
  } else {
    probeAvailability();
  }
})();
